lib/OpenQA/Scheduler/Scheduler.pm

```diff
@@ -138,17 +150,397 @@ sub _prefer_parallel {
+                log_debug("[Job#${j}] Accepted by worker $expected_workerid - setted to running state");
+            }
+            catch {
+                # Aborts and set the job to scheduled again
+                $job->reschedule_rollback if $job->result eq OpenQA::Schema::Result::Jobs::NONE;
+                $failure++
+                  if OpenQA::Scheduler::CONGESTION_CONTROL()
+                  && OpenQA::Scheduler::BUSY_BACKOFF();
+            };
+        }
+        else {
+            log_debug("[Job#${j}] Too bad, seems wasn't accepted - sending abort to worker");
+
+            $job->reschedule_rollback;
+
+            $failure++
+              if OpenQA::Scheduler::CONGESTION_CONTROL()
+              && OpenQA::Scheduler::BUSY_BACKOFF();    # double it, we might be in a heavy load condition.
+                                                       # Also, the workers could be massively in re-registration.
+        }
+    }
+    $ws_allocated_jobs = {};
+
+    # Avoid to go into starvation - reset the scheduler tick counter.
+    reactor->{timer}->{capture_loop_avoidance} ||= reactor->add_timeout(
+        OpenQA::Scheduler::CAPTURE_LOOP_AVOIDANCE(),
+        Net::DBus::Callback->new(
+            method => sub {
+                return if $failure == 0;
+                $failure = 0;
+                log_debug("[Congestion control] Resetting failures count. Next scheduler round will be reset to "
+                      . OpenQA::Scheduler::SCHEDULE_TICK_MS()
+                      . "ms");
+                _reschedule(
+                    OpenQA::Scheduler::BUSY_BACKOFF() ?
+                      OpenQA::Scheduler::SCHEDULE_TICK_MS() + 1000
+                    : OpenQA::Scheduler::SCHEDULE_TICK_MS());
+            })) if (OpenQA::Scheduler::CONGESTION_CONTROL());
+
+    my @allocated_jobs;
+
+    log_debug("-> Scheduling new jobs.");
+    try {
+        @allocated_jobs = schema->txn_do(
+            sub {
+                my $all_workers = schema->resultset("Workers")->count();
+                my @allocated_workers;
+                my @allocated_jobs;
+
+                # $worker->connected is too much expensive, prefeer dead.
+                my @free_workers = grep { !$_->dead } schema->resultset("Workers")->search({job_id => undef},)->all();
+                @free_workers = shuffle(@free_workers)
+                  if OpenQA::Scheduler::SHUFFLE_WORKERS();   # shuffle avoids starvation if a free worker keeps failing.
+                log_debug("\t Free workers: " . scalar(@free_workers) . "/$all_workers");
+                log_debug("\t Failure# ${failure}") if OpenQA::Scheduler::CONGESTION_CONTROL();
+
+                # Consider it a failure if OpenQA::Scheduler::CONGESTION_CONTROL is set
+                # so if there are no free workers scheduler will kick in later.
+                $failure++
+                  and return ()
+                  if @free_workers == 0
```

Comment:
Note: to be more performant, we _should_ return even if `CONGESTION_CONTROL` and `BUSY_BACKOFF` are disabled